asim: fix division by zero in findReplicaPlacementForEveryStoreSet

wenyihu6 opened the pull request:
**asim: fix division by zero in findReplicaPlacementForEveryStoreSet**

We previously enabled multiple simulation runs within the same datadriven file
to compare different configurations under a shared test setup. To support this,
we began reusing the same rangeGen across configurations. However,
state.ReplicaPlacement was mutable, and findReplicaPlacementForEveryStoreSet
updated the weights to reflect actual range counts. These weights were then
decremented during RangesInfoWithReplicaPlacement to track remaining
allocations, eventually reaching zero. This commit fixes the issue by making
state.ReplicaPlacement immutable and introducing a wrapper to manage range count
computation and allocation separately.

Because the weights were zero during subsequent configuration runs,
findReplicaPlacementForEveryStoreSet would divide by zero due to a zero total
weight. This caused architecture dependent behavior, leading to test failures on
TeamCity/gce-workers that couldn’t be reproduced locally on mac.

Fixes: #151263
Epic: none

---

**asim: revert "asim: revert example_skewed_cpu_even_ranges_mma back to mma-only"**

Previously, we restricted example_skewed_cpu_even_ranges_mma to run only with
mma-only mode due to CI failures. Now that the underlying bug has been
identified and fixed, we can re-enable it for all configurations.

Epic: none
Release note: none

wenyihu6 committed:
asim: fix division by zero in findReplicaPlacementForEveryStoreSet

We previously enabled multiple simulation runs within the same datadriven file
to compare different configurations under a shared test setup. To support this,
we began reusing the same rangeGen across configurations. However,
state.ReplicaPlacement was mutable, and findReplicaPlacementForEveryStoreSet
updated the weights to reflect actual range counts. These weights were then
decremented during RangesInfoWithReplicaPlacement to track remaining
allocations, eventually reaching zero. This commit fixes the issue by making
state.ReplicaPlacement immutable and introducing a wrapper to manage range count
computation and allocation separately.

Because the weights were zero during subsequent configuration runs,
findReplicaPlacementForEveryStoreSet would divide by zero due to a zero total
weight. This caused architecture dependent behavior, leading to test failures on
TeamCity/gce-workers that couldn’t be reproduced locally on mac.

Fixes: #151263
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151421)
<!-- Reviewable:end -->

wenyihu6 committed:
asim: revert "asim: revert example_skewed_cpu_even_ranges_mma back to mma-only"

Previously, we restricted example_skewed_cpu_even_ranges_mma to run only with
mma-only mode due to CI failures. Now that the underlying bug has been
identified and fixed, we can re-enable it for all configurations.

Epic: none
Release note: none

wenyihu6 commented:
Rebased this pr on top of master to revert commit da2c7c21b28f. This likely makes reviewable unpleasant, but hoping this pr is small enough.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151421#-:-OWzzAOPE2nTPSCiDHvn:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->
